arraycontext/container/traversal.py

```diff
@@ -279,6 +295,52 @@ def rec(keys: Tuple[Union[str, int], ...],
 # }}}
 
 
+# {{{ array container reductions
+
+def rec_reduce_array_container(
+        reduce_func: Callable[[Iterable[Any]], Any],
+        array_func: Callable[[Any], Any],
+        ary: ArrayContainerT) -> Any:
+    """Perform reductions over array containers recursively.
+
+    :param reduce_func: callable used to reduce over the components of the
+        :class:`~arraycontext.ArrayContainer`.
+    :param array_func: callable used to reduce a single component of the
+        :class:`~arraycontext.ArrayContainer`. The callable takes arrays of
+        type :class:`arraycontext.ArrayContext.array_types` and returns an
+        array of the same type or a scalar.
+    """
+    def _reduce_wrapper(ary: Any, iterable: Iterable[Tuple[Any, Any]]) -> Any:
```

Comment:
Why not avoid this wrapper altogether? (And just change the prototype of `reduce_func` to match?)

Reply:
Not really a fan because it pushes an implementation detail on the user. Although maybe there is some fancy reduction that would need to know about the `template_ary` and keys being reduced?

This is only used in the `multimap` case now. Do you still think it should be changed?

Reply:
Alright with me if it's `multimap`-only.